Thread-local default_random_engine broken on Clang/Mac OS

Commit aa62c9f causes a compile error with Clang (AppleClang 7.3.0) on Mac OS when the utility/random.hpp header is included:

```
range-v3/include/range/v3/utility/random.hpp:500:66: error: initializer for thread-local variable must be a constant expression
                RANGES_STATIC_THREAD_LOCAL default_random_engine engine{
                                                                 ^~~~~~~
```

A workaround is to simply change the definition of RANGES_STATIC_THREAD_LOCAL to just `static` (rather than `static __thread`) on Apple systems, but of course this is less than ideal...

--- a/include/range/v3/detail/config.hpp
+++ b/include/range/v3/detail/config.hpp
@@ -58,16 +58,6 @@
 #define RANGES_END_NAMESPACE_STD }
 #endif
 
-#ifndef RANGES_STATIC_THREAD_LOCAL
-#if defined(__clang__) && (defined(__CYGWIN__) || defined(__apple_build_version__))
-// BUGBUG avoid unresolved __cxa_thread_atexit
-// Also, workaround lack of thread_local support on OSX
-#define RANGES_STATIC_THREAD_LOCAL static __thread
-#else
-#define RANGES_STATIC_THREAD_LOCAL static thread_local
-#endif
-#endif
-
 // Database of feature versions
 #define RANGES_CXX_STATIC_ASSERT_11 200410
 #define RANGES_CXX_STATIC_ASSERT_14 RANGES_CXX_STATIC_ASSERT_11
@@ -87,6 +77,10 @@
 #define RANGES_CXX_RETURN_TYPE_DEDUCTION_14 201304
 #define RANGES_CXX_GENERIC_LAMBDAS_11 0
 #define RANGES_CXX_GENERIC_LAMBDAS_14 201304
+#define RANGES_CXX_STD_11 201103
+#define RANGES_CXX_STD_14 201402
+#define RANGES_CXX_THREAD_LOCAL_11 RANGES_CXX_STD_11
+#define RANGES_CXX_THREAD_LOCAL_14 RANGES_CXX_THREAD_LOCAL_11
 
 #if defined(_MSC_VER) && !defined(__clang__)
 #if _MSC_VER >= 1900
@@ -168,8 +162,6 @@
 #define RANGES_CXX_FEATURE_CONCAT(y, z) RANGES_CXX_FEATURE_CONCAT2(y, z)
 #define RANGES_CXX_FEATURE(x) RANGES_CXX_FEATURE_CONCAT(x, RANGES_CXX_STD_NAME)
 
-#define RANGES_CXX_STD_11 201103
-#define RANGES_CXX_STD_14 201402
 #if __cplusplus >= RANGES_CXX_STD_14
 #define RANGES_CXX_STD_NAME 14
 #define RANGES_CXX_STD RANGES_CXX_STD_14
@@ -245,6 +237,16 @@
 #endif
 #endif
 #endif // MSVC/Generic configuration switch
+
+#ifndef RANGES_CXX_THREAD_LOCAL
+#if defined(__clang__) && (defined(__CYGWIN__) || defined(__apple_build_version__))
+// BUGBUG avoid unresolved __cxa_thread_atexit
+// Also, workaround lack of thread_local support on OSX
+#define RANGES_CXX_THREAD_LOCAL 0
+#else
+#define RANGES_CXX_THREAD_LOCAL RANGES_CXX_FEATURE(THREAD_LOCAL)
+#endif
+#endif
 
 #ifndef RANGES_DISABLE_DEPRECATED_WARNINGS
 #if RANGES_CXX_ATTRIBUTE_DEPRECATED && \

--- a/include/range/v3/utility/random.hpp
+++ b/include/range/v3/utility/random.hpp
@@ -51,6 +51,7 @@
 #include <chrono>
 #include <functional>  // for std::hash
 #include <initializer_list>
+#include <new>
 #include <random>
 #include <thread>
 #include <typeinfo>
@@ -497,9 +498,23 @@ namespace ranges
                 meta::if_c<sizeof(void*) >= 8, std::mt19937_64, std::mt19937>;
             inline default_random_engine& get_random_engine()
             {
-                RANGES_STATIC_THREAD_LOCAL default_random_engine engine{
+#if RANGES_CXX_THREAD_LOCAL
+                static thread_local default_random_engine engine{
                     randutils::auto_seed_256{}.base()};
                 return engine;
+#else
+                static __thread bool initialized = false;
+                static __thread meta::_t<std::aligned_storage<
+                    sizeof(default_random_engine),
+                    alignof(default_random_engine)>> storage;
+
+                if (!initialized) {
+                    ::new(static_cast<void*>(&storage)) default_random_engine{
+                        randutils::auto_seed_256{}.base()};
+                    initialized = true;
+                }
+                return reinterpret_cast<default_random_engine&>(storage);
+#endif
             }
         }
         /// \endcond

